README.md
# PostCSS Image Inliner [![Build Status][ci-img]][ci] [![Windows Build Status][winci-img]][winci]

[PostCSS] plugin to inline local/remote images.

[PostCSS]: https://github.com/postcss/postcss
[ci-img]:  https://travis-ci.org/bezoerb/postcss-image-inliner.svg
[ci]:      https://travis-ci.org/bezoerb/postcss-image-inliner
[winci-img]: https://ci.appveyor.com/api/projects/status/5xaq0ord84y5ho0c?svg=true
[winci]: https://ci.appveyor.com/project/bezoerb/postcss-image-inliner

```css
.foo {
    /* Input example */
    background-image: url("https://placehold.it/10x10");
}
```

```css
.foo {
  /* Output example */
  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAo ... ORK5CYII=");
}
```

## Usage

```js
var postcss = require('postcss');
var imageInliner = require('postcss-image-inliner');
var opts = {
    assetPaths: [],     // List of directories where the inliner should look for assets
    maxFileSize: 10240  // Sets a max file size (in bytes)
}

postcss([ imageInliner(opts) ])
```

See [PostCSS] docs for examples for your environment.

## Options
#### assetPaths
Type: `array`<br/>
Default: `[process.cwd()]`<br/>
Example: `['http://domain.de/', 'http://domain.de/styles', 'app/images', '**/images/']`<br/>
Required: `false`<br/>

List of directories/urls where the inliner should start looking for assets.
You can define local directories (globs supported) or URLs.


#### maxFileSize
Type: `int`<br/>
Default: `10240`<br/>
Example: `0`<br/>
Required: `false`<br/>

Sets a max file size (in bytes) for inlined images. Set to `0` to disable size checking.

#### b64Svg
Type: `bool`<br/>
Default: `false`<br/>
Required: `false`<br/>

Use Base64 encoding for svgs.

#### strict
Type: `bool`<br/>
Default: `false`<br/>
Required: `false`<br/>

Fail on error


CHANGELOG.md

v3.0.1 / 2019-06-14
===================

  * Test node 12
  * Bump dependencies

v3.0.0 / 2019-01-19
===================

  * Rewrite
  * Drop node support < 8

v2.0.0 / 2017-12-17
===================

  * Drop support for node < 6
  * Bump dependencies
  * Add strict mode

v1.0.6 / 2017-02-26
==================

  * Bump dependencies
  * Add peer dependency

v1.0.5 / 2017-02-26
===================

  * bump deps

v1.0.4 / 2017-01-30
===================

  * bump deps
  * Fix the issue with "cursor" property (#156)

v1.0.3 / 2017-01-29
===================

  * bump deps

v1.0.2 / 2016-11-22
===================

  * removed unnecessary escape character
  * bump deps

v1.0.1 / 2016-09-28
===================

  * bump deps

v1.0.0 / 2016-09-02
===================

  * Minimum node.js version is 4.0 now due to upstream changes.
  * bump deps

v0.3.1 / 2016-05-10
===================

  * travis tweaks
  * bump deps & eslint-config
  * added glob tests

v0.3.0 / 2016-02-05
===================

  * handle content property (#61)
  * bump deps

0.2.0 / 2015-12-10
==================

  * Added 'strict' option (#28)
  * Bump dependencies

0.1.3 / 2015-09-24
==================

  * Better svg encoding + b64Svg option (#1)

0.1.2 / 2015-09-24
==================

  * Outsourced asset-resolver

0.1.1 / 2015-09-20
==================

  * Fixed filesize check

0.1.0 / 2015-09-19
==================

  * Initial release


test/test.js
/* eslint-env node, mocha */
const path = require('path');
const fs = require('fs');
const http = require('http');
const postcss = require('postcss');
const assert = require('chai').assert;
const finalhandler = require('finalhandler');
const serveStatic = require('serve-static');

const plugin = require('../');

function read(filename) {
  return fs.readFileSync(path.join(__dirname, 'fixtures', 'styles', filename), 'utf8');
}

const test = function (input, output, opts = {}, done) {
  input = read(input);
  output = read(output);

  const options = {assetPaths: ['//localhost:3000/styles/'], ...opts};

  postcss([plugin(options)])
    .process(input, {from: undefined})
    .then(result => {
      assert.equal(result.css, output);
      assert.isEmpty(result.warnings());
    })
    .then(done)
    .catch(err => {
      done(err);
    });
};

function startServer(docroot) {
  const serve = serveStatic(docroot);
  const server = http.createServer((req, res) => {
    const done = finalhandler(req, res);
    serve(req, res, done);
  });
  server.listen(3000);

  return server;
}

/* eslint max-len:0 */
describe('postcss-image-inliner', () => {
  let server;

  beforeEach(() => {
    server = startServer(path.join(__dirname, 'fixtures'));
  });
  afterEach(done => {
    server.close(done);
  });

  it('should skip too big images', done => {
    test('big.css', 'big.css', {maxFileSize: 1}, done);
  });

  it('should handle multiple background images', done => {
    test('multi.in.css', 'multi.out.css', {}, done);
  });

  it('should handle background shorthand', done => {
    test('shorthand.in.css', 'shorthand.out.css', {}, done);
  });

  it('should handle media queries', done => {
    test('media.in.css', 'media.out.css', {}, done);
  });

  it('should consider base64Svg option', done => {
    test('b64svg.in.css', 'b64svg.out.css', {b64Svg: true, maxFileSize: 0}, done);
  });

  it('should do nothing on missing files in non strict mode', done => {
    test('missing.in.css', 'missing.out.css', {}, done);
  });

  it('should handle urls used for content property', done => {
    test('pseudo.in.css', 'pseudo.out.css', {}, done);
  });

  it('should fail on missing files in strict mode', done => {
    test('missing.in.css', 'missing.out.css', {strict: true}, error => {
      if (error) {
        done();
      } else {
        done(new Error('Should fail'));
      }
    });
  });

  it('should allow globbing', done => {
    test('glob.in.css', 'glob.out.css', {assetPaths: 'test/*/images/'}, done);
  });
});


index.js
const postcss = require('postcss');
const debug = require('debug')('image-inliner');
const last = require('lodash/last');
const escapeRegExp = require('lodash/escapeRegExp');
const {getDataUriMapping} = require('./lib/image');

const DEFAULTS = {
  assetPaths: [],
  maxFileSize: 10240,
  b64Svg: false,
  strict: false
};

const loop = cb => {
  const matcher = /url\(\s*(?:['"]*)(?!['"]*data:)(.*?)(?:['"]*)\s*\)/gm;
  return decl => {
    let match;
    while ((match = matcher.exec(decl.value)) !== null) {
      cb({decl, url: last(match)});
    }
  };
};

module.exports = postcss.plugin('postcss-image-inliner', (opts = {}) => {
  const options = {...DEFAULTS, ...opts};

  if (Array.isArray(options.assetPaths)) {
    options.assetPaths = [...options.assetPaths, process.cwd()];
  } else {
    options.assetPaths = [options.assetPaths, process.cwd()];
  }

  return async css => {
    const urls = new Set([]);
    const filter = /^(background(?:-image)?)|(content)|(cursor)/;
    // Get urls
    css.walkDecls(filter, loop(({url}) => urls.add(url)));

    const mapping = await getDataUriMapping([...urls], options);

    css.walkDecls(
      filter,
      loop(({decl, url}) => {
        if (mapping[url]) {
          const regexp = new RegExp('[\'"]?' + escapeRegExp(url) + '[\'"]?', 'gm');
          if (url.startsWith('//')) {
            //       Console.log({url, regexp, val: decl.value});
          }

          decl.value = decl.value.replace(regexp, '\'' + mapping[url] + '\'');
          if (url.startsWith('//')) {
            //       Console.log('RESULT: ', decl.value);
          }

          debug(url, 'successfully replaced with ', mapping[url]);
        } else {
          debug(url, 'failed');
          if (opts.strict) {
            throw new Error('No file found for ' + url);
          }
        }
      })
    );
  };
});


lib/resource.js
/**
 * Created by ben on 17.09.15.
 */
var fs = require('fs');
var url = require('url');
var request = require('request');
var mime = require('mime');
var Promise = require('bluebird');
var path = require('path');
var debug = require('debug')('image-inliner');
var result = require('lodash.result');
var filesize = require('filesize');
var hash = require('object-hash');
var cache = {};

Promise.promisifyAll(fs);

function getSize(string, binary) {
    return Buffer.byteLength(string, binary ? 'binary' : 'utf-8');
}

/**
 * Get external resource
 * @param {string} uri
 * @returns {Promise}
 */
function requestAsync(resource, opts) {
    var maxFileSize = result(opts, 'maxFileSize');
    var settings = {
        followRedirect: true,
        encoding:       'binary'
    };
    return new Promise(function (resolve, reject) {
        // handle protocol-relative urls
        resource = url.resolve('http://te.st', resource);
        request(resource, settings, function (err, resp, body) {
            var msg;
            if (err) {
                debug('Url failed:', err.message || err);
                return reject(err);
            }
            if (resp.statusCode !== 200) {
                msg = 'Wrong status code ' + resp.statusCode + ' for ' + resource;
                debug(msg);
                return reject(msg);
            }

            var mimeType = result(resp, 'headers.content-type') || mime.lookup(resource);
            var size = getSize(body, mimeType !== 'image/svg+xml');
            if (maxFileSize && maxFileSize < size) {
                msg = 'Too big.  ' + filesize(size) + ' exceeds the allowed ' + filesize(maxFileSize);
                debug(msg);
                return reject(msg);
            }

            debug('Fetched:', resource);
            resolve({
                contents: body,
                path:     resource,
                mime:     mimeType
            });
        });
    });
}

function isUrl(resource) {
    return /(^\/\/)|(:\/\/)/.test(resource);
}


function readAsync(resource, opts) {
    var maxFileSize = result(opts, 'maxFileSize');
    return fs.readFileAsync(resource).then(function (body) {
        var mimeType = mime.lookup(resource);
        var size = getSize(body, mimeType !== 'image/svg+xml');
        if (maxFileSize && maxFileSize < size) {
            var msg = 'Too big.  ' + filesize(size) + ' exceeds the allowed ' + filesize(maxFileSize);
            debug(msg);
            return new Error(msg);
        }
        debug('Fetched:', resource);
        return {
            contents: body,
            path:     resource,
            mime:     mimeType
        };
    });
}

function join(base, file) {
    if (isUrl(file)) {
        return file;
    }
    if (isUrl(base)) {

        return url.resolve(base, file);
    }

    return path.join(base, file);
}

function getResource(base, file, opts) {
    var maxFileSize = result(opts, 'maxFileSize') || 0;
    var resource = join(base, file);
    var cacheKey = hash({
        file: resource,
        size: maxFileSize
    });

    if (cache[cacheKey]) {
        return cache[cacheKey];
    }

    if (isUrl(resource)) {
        cache[cacheKey] = requestAsync(resource, opts);
    } else {
        cache[cacheKey] = readAsync(resource, opts);
    }
    return cache[cacheKey];
}


module.exports.getResource = getResource;


lib/image.js
'use strict';
const path = require('path');
const SVGO = require('svgo');
const filesize = require('filesize');
const debug = require('debug')('image-inliner');
const {getResource} = require('asset-resolver');

const svgo = new SVGO();

// Borrowed from https://github.com/filamentgroup/directory-encoder/blob/master/lib/svg-uri-encoder.js
function encodeSvg(content) {
  return (
    encodeURIComponent(
      content
        // Strip newlines and tabs
        .replace(/[\n\r]/gim, '')
        .replace(/\t/gim, ' ')
        // Strip comments
        .replace(/<!--(.*(?=-->))-->/gim, '')
        // Replace
        .replace(/'/gim, '\\i')
    )
      // Encode brackets
      .replace(/\(/g, '%28')
      .replace(/\)/g, '%29')
      // Replace ' with "
      .replace(/'/gm, '"')
  );
}

async function reduceAsync(array = [], reducer = r => r, initial) {
  for (const index of array.keys()) {
    initial = await reducer(initial, array[index], index); /* eslint-disable-line no-await-in-loop */
  }

  return initial;
}

async function assertSize(resource, maxFileSize) {
  const {mime, contents = ''} = resource || {};
  const encoding = mime === 'image/svg+xml' ? 'utf-8' : 'binary';
  const size = Buffer.byteLength(contents, encoding);
  if (maxFileSize && maxFileSize < size) {
    const msg = 'Too big.  ' + filesize(size) + ' exceeds the allowed ' + filesize(maxFileSize);
    throw new Error(msg);
  }

  return resource;
}

async function getDataUri(file, opts) {
  // If the url is SVG, let's compress and use the XML directly
  if (file.mime === 'image/svg+xml' && !opts.b64Svg) {
    const optimized = await svgo.optimize(file.contents.toString('utf-8'));
    return `data:image/svg+xml;charset=US-ASCII,${encodeSvg(optimized.data)}`;
  }

  // Otherwise we base64 encode the image
  return 'data:' + file.mime + ';base64,' + Buffer.from(file.contents, 'binary').toString('base64');
}

async function resolve(filepath, {assetPaths, maxFileSize, resolver = () => undefined}) {
  const filter = resource => assertSize(resource, maxFileSize);
  const resolvedFile = await resolver(filepath);
  const base = resolvedFile ? path.dirname(resolvedFile) : assetPaths;
  const file = resolvedFile ? path.basename(resolvedFile) : filepath;

  try {
    return await getResource(file, {base, filter});
  } catch (err) {
    debug(err.message, filepath, 'could not be resolved');
  }
}

function getDataUriMapping(urls = [], options = {}) {
  return reduceAsync(
    [...new Set(urls)],
    async (result, url) => {
      const file = await resolve(url, options);
      if (file && file.mime && /image/.test(file.mime)) {
        result[url] = await getDataUri(file, options);
      }

      return result;
    },
    {}
  );
}

module.exports.getDataUriMapping = getDataUriMapping;
